src/events/SDL_events.c

```diff
@@ -785,10 +788,14 @@ SDL_PollEvent(SDL_Event * event)
 static int
 SDL_WaitEventTimeout_Device(_THIS, SDL_Window *wakeup_window, SDL_Event * event, int timeout)
 {
-    /* Release any keys held down from last frame */
-    SDL_ReleaseAutoReleaseKeys();
-
     for (;;) {
+        /* Pump events on entry and each time we wake to ensure:
+           a) All pending events are batch processed after waking up from a wait
+           b) Waiting can be completely skipped if events are already available to be pumped
+           c) Periodic processing that takes place in some platform PumpEvents() functions happens
+        */
+        SDL_PumpEvents();
```

Comment:
I thought the call to PumpEvents should be placed after we acquired the wakeup_lock.

In addition we may just call `_this->PumpEvents(_this)`, it makes more sense to me since we already have the _this video device but I am not completely sure. I don't know if we need the call to `SDL_SendPendingSignalEvents()` which is otherwise called in SDL_PumpEvents.

Reply:
What's the thinking behind needing to pump events under the lock?

IIUC, the lock is protecting us from the possible race where we'd miss a wakeup if a `SDL_SendWakeupEvent()` call happens after `SDL_PeepEvents()` returns 0, but before setting `_this->wakeup_window` to our window. Any events that were pushed by `SDL_PumpEvents()` outside the lock would be returned by the `SDL_PeepEvents()` call after we acquire the lock and before we wait, so there's no risk of missing a wakeup there.

Reply:
I think you are right, there is no reason for the PumpEvents to go inside the lock.

In fact I think that the PumpEvents call you added is not required at all but it is just an opportunist optimization to catch a last-moment event just before we may block waiting for an event.

Reply:
It's not just an opportunistic optimization. There are real breakages today due to the lack of the `PumpEvents()` call (see `c)` in my comment).

Holding both Left and Right Shift on Windows then releasing both is broken if you exclusively use `SDL_WaitEvent()` today.

This is because the corrective code for this case only runs in `PumpEvents()`.
https://github.com/libsdl-org/SDL/blob/41fab8532d56f8b286195edbc2d9bf270fbe5ca0/src/video/windows/SDL_windowsevents.c#L1355-L1365

Reply:
Thank you for the details, now the problem is more clear to me.

I suggest that in the Windows video device, in the function WIN_WaitEventTimeout we do not call the blocking GetMessage function if one of the shift key is currently pressed. Instead we poll and we perform the "heroic" workaround to sense a shift KEYUP event that windows doesn't send.

It seems to me like a more robust solution and in addition it will affect only the WIN backend which seems more correct since it is a window's specific problem.

If you agree on this approach we may remove the PumpEvents call you are adding in this PR.

For the other problem, the call to WIN_UpdateClipCursorForWindows(), I don't have a solution to propose for the moment.

Reply:
What do we gain by avoiding `PumpEvents()`?

Processing window messages one at a time (as `WaitEventTimeout()` does) means extra work for _every_ window message (including those that don't even result in an SDL event). We have to do this dance of setting and killing a Win32 timer (for waits with a timeout) for every window message. With high polling rate gaming mice in raw input mode, that could be thousands of messages per second.

I also worry about the consequences of constraining us to process only one window message per `SDL_WaitEventTimeout()` call. If user activity is generating more than one window message per application `SDL_WaitEventTimeout()` call, we could be left with a growing message queue backlog that may not recover.

Reply:
> What do we gain by avoiding `PumpEvents()`?

I thought it doesn't really solve the problem and it is just confusing but I may be wrong.

Why it doesn't solve the problem ? For the shift KEYUP problem you have to continuously probe the keyboard state to see if one shift was released because windows is not sending the event when you release one of the shift key. So you need to probe the system all the time because windows doesn't send the event. Now if you block waiting for an event that means that you cannot probe the keyboard status because you are blocking and it doesn't matter if you called PumpEvents before.

Am I missing something ?

> Processing window messages one at a time (as `WaitEventTimeout()` does) means extra work for _every_ window message (including those that don't even result in an SDL event). We have to do this dance of setting and killing a Win32 timer (for waits with a timeout) for every window message. With high polling rate gaming mice in raw input mode, that could be thousands of messages per second.

But processing one event at a time is in the semantic of WaitEventTimeout. Normally I consider that the best practice for applications using WaitEvent or WaitEventTimeout is that, when an event occurs, a loop should immediately follow to poll for events to quickly empty the event queue and treat a burst of events as quickly as possible.

Otherwise I am afraid like you of the overhead of setting a Timer each time but I don't see any other solution. On the other side it seems reasonable to do this if the developer explicitly requested a timeout since on Windows this is what you are supposed to do.

May be we could add an option in the SDL library to request actively polling behavior for WaitEvent / WaitEventTimeout functions. In this way we may defuse the tension we have now between games and desktop applications. Games could request a polling behavior for WaitEvent / WaitEventTimeout to avoid all kind of problems.

Reply:
There are a couple problems that we're talking about here which is probably what's causing some of the confusion.
1. There are special actions (like tickling the screensaver, handling both shift key presses, updating the cursor clip region, etc) which need to be performed periodically that aren't happening at all with the current code.
2. Handling just a single window message at a time causes additional overhead that isn't necessary and could possibly lead to a backlog of window messages in some cases.

Calling `SDL_PumpEvents()` in the way I've done here partially solves problem 1 and completely solves problem 2. I agree more changes are required to completely solve problem 1, like those we discussed earlier.

> But processing one event at a time is in the semantic of WaitEventTimeout.
That's true, but I'm talking about window messages not SDL events. There is not a 1 to 1 relationship between window messages and SDL events. Many types of window message aren't handled by SDL at all, while others are only used to update internal state which doesn't end up generating an SDL event.

The semantics `SDL_WaitEventTimeout()` and `SDL_PollEvent()` also don't preclude doing batch event processing inside SDL, and indeed `SDL_PollEvent()` does do that via `SDL_PumpEvents()` like I am adding here.

> Otherwise I am afraid like you of the overhead of setting a Timer each time but I don't see any other solution. On the other side it seems reasonable to do this if the developer explicitly requested a timeout since on Windows this is what you are supposed to do.

I don't think we're breaking the API contract by calling `SDL_PumpEvents()` inside `SDL_WaitEventTimeout()`. This is exactly what we used to do before the new wait code was introduced (and what we still do for video backends that don't support `WaitEventTimeout()`). The timeout still works as intended with or without pumping events.

> May be we could add an option in the SDL library to request actively polling behavior for WaitEvent / WaitEventTimeout functions. In this way we may defuse the tension we have now between games and desktop applications. Games could request a polling behavior for WaitEvent / WaitEventTimeout to avoid all kind of problems.

We still wait for new SDL events as before, only waking up when there are new window messages or SDL events to process. The call to `SDL_PumpEvents()` simply allows us to batch process the pending window messages when we wake up from the wait or avoid the wait altogether if events are available prior to waiting. I believe this behavior is desirable for both games and desktop applications.